### PR TITLE
Standardize repository runtime on Python 3.13

### PR DESCRIPTION
**Pull request by tnorlund**
## Summary

- Standardize all active package metadata, developer tooling, and CI on Python 3.13.
- Upgrade Lambda runtimes, Lambda container bases, layer compatibility, SageMaker launch code, and the EMR Serverless image to Python 3.13.
- Add `.python-version` plus a repository consistency check so active Python targets cannot silently drift apart again.
- Raise `receipt_layoutlm`'s scikit-learn floor to 1.6, the first release line with Python 3.13 wheels.

## Why

The repository had accumulated Python targets from 3.8 through 3.12 across packages, CI, deployment definitions, containers, and runner scripts. That made local and deployed environments inconsistent and allowed future updates to cover only part of the stack.

This PR establishes Python 3.13 as the single active baseline and enforces it in CI.

## Impact

- All 12 active Python packages now require Python 3.13.
- 38 Lambda runtime declarations, 19 Lambda base images, and 5 Lambda layer compatibility declarations move to 3.13.
- `receipt_chroma` remains capped at `<3.14` because of its current dependency constraints.
- Existing `py312` physical resource identifiers in the legacy receipt-label cleanup script are intentionally preserved; renaming deployed resources would be a separate migration.
- Merging this changes production deployment targets, so the PR remains draft pending GitHub Actions and infrastructure review.

## Validation

- Python version consistency check, `pip check`, `compileall`, Black, and isort pass under Python 3.13.12.
- Repository tests: 772 passed, 1 skipped.
- `receipt_dynamo`: 3,948 passed.
- `receipt_dynamo_stream`: 150 passed.
- `receipt_chroma`: 571 passed, 7 skipped.
- `receipt_places`: 169 passed.
- `receipt_langsmith`: 76 passed, 40 skipped.
- `receipt_upload`: full suite passed.
- `receipt_agent`: 387 passed, 22 skipped.
- `receipt_layoutlm`: 121 passed with the CI marker exclusions.
- `receipt_logo`: 10 passed.
- Glyph Python suite: 286 passed.
- Infrastructure smoke tests: 21 passed.
- Every package manifest builds successfully, including the ChromaDB compaction wheel.
- Lambda Python 3.13 x86_64 and arm64 container manifests resolve successfully.
- The generated EMR Dockerfile uses Python 3.13 and avoids replacing the AL2023 system `python3` link.

## Known unrelated test debt

The legacy ChromaDB compaction Lambda test directory is not part of current CI and still fails when invoked directly because of stale `StreamMessage` fixtures, ambiguous `utils` imports, and Pulumi import side effects. Its runtime declarations were updated here, but its source and test architecture were otherwise left unchanged.

### CODE REVIEW

File: receipt_layoutlm/pyproject.toml [hunk 1]
```diff
@@ -35,7 +36,8 @@ dependencies = [
   "Pillow>=10.0.0,<13.0.0",
   "sentencepiece>=0.1.99,<1.0.0",
   "seqeval>=1.2.2,<2.0.0",
-  "scikit-learn>=1.3.0,<=1.5.1",
+  # 1.6 is the first release with CPython 3.13 wheels.
+  "scikit-learn>=1.6.0,<2.0.0",
```
[the remark this answers is not in the document]
Comment (tnorlund) [reply receipt_layoutlm/pyproject.toml]: Fixed in `66df29dca` by separating the incompatible environments instead of forcing one scikit-learn range across both:

- `receipt_layoutlm[training]` now owns scikit-learn `>=1.6,<2` and `seqeval` for Python 3.13 training metrics.
- `receipt_layoutlm[coreml]` pins CoreMLtools 9.0, excludes scikit-learn, and the shared Torch range is capped at CoreMLtools' tested `<=2.7.0` ceiling.
- The CoreML worker runbook now installs the `coreml` extra, while SageMaker explicitly installs the `training` extra.
- Added a dependency-contract regression test for this split.

End-to-end validation used a clean Python 3.13.12 macOS ARM venv: CoreMLtools 9.0 + Torch 2.7.0 resolved with no scikit-learn, `pip check` passed, the CLI imported and started without unsupported-version warnings, and a real `LayoutLMWrapper` was traced, converted to an ML Program, executed through CoreML, and matched PyTorch in FLOAT32 with maximum absolute error `2.24e-8`.

One separate pre-existing risk remains explicit: the synthetic wrapper produced NaNs through the existing FLOAT16 path. I did not change production's quantization default without a real-checkpoint parity test; the Python 3.13/CoreML dependency conflict raised here is resolved and the validated FLOAT32 path is sound.

File: scripts/check_python_version_consistency.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    ".github",
+    ".codex/skills",
+    "infra",
+    "scripts",
+    "synthesis_loop",
+    "tests",
+    "receipt_langsmith/scripts",
+    "receipt_upload/tests",
+)
+SCAN_SUFFIXES = {".py", ".sh", ".yaml", ".yml"}
```
[the remark this answers is not in the document]
Comment (tnorlund) [reply scripts/check_python_version_consistency.py:24]: Fixed in `66df29dca`:

- The consistency checker now scans all maintained Markdown for old Python declarations and prose references.
- Historical evidence is explicitly excluded only under `.review-loop`, `docs/archive`, `docs/handoff`, `docs/handoffs`, and `docs/review-loops`; generated environments/dependency trees are also excluded.
- The cited `infra/chromadb_compaction/tests/TESTING_GUIDE.md` workflow now uses Python 3.13.
- The broader scan found and corrected seven other maintained documents that still said Python 3.8 or 3.12.
- Added regression cases proving maintained Markdown fails the check and every historical root remains excluded.

Validation: the checker reports `All active Python targets use Python 3.13`; the combined review regression suite is 16/16 green, and the repository suite is 761 passed / 1 skipped.

File: receipt_layoutlm/pyproject.toml [hunk 3]
```diff
@@ -7,12 +7,13 @@ name = "receipt_layoutlm"
 version = "0.1.0"
 description = "LayoutLM training package for receipt understanding integrated with DynamoDB"
 readme = "README.md"
-requires-python = ">=3.10"
+requires-python = ">=3.13"
```
[the remark this answers is not in the document]
Comment (tnorlund) [reply receipt_layoutlm/pyproject.toml:10]: Fixed in `66df29dca` while preserving the proven PyTorch 2.6/CUDA 12.4 base image:

- The image creates and activates an isolated Conda Python 3.13 environment before either local package is installed.
- It installs the official CPython 3.13 `torch 2.6.0+cu124` wheel and asserts Python minor, Torch version, and CUDA ABI at build time.
- SageMaker now installs `receipt_layoutlm[training]`, runs `pip check`, and imports the complete training stack during the image build.
- Added regression tests for activation/install order, the CUDA/Torch contract, and epoch-eval's reuse of this same training image.

End-to-end validation: a real `linux/amd64` Docker build of the dependency stage succeeded, and I ran the resulting container. It reports Python 3.13.15, Torch 2.6.0+cu124 with CUDA 12.4, scikit-learn 1.9.0, and seqeval 1.2.2; both package imports and `pip check` pass. Epoch evaluation directly reuses this corrected image through `ECR_IMAGE_URI`.

The Mac cannot execute NVIDIA kernels, so GPU kernel execution remains an AWS runtime check, but the interpreter/package incompatibility that prevented the image from building is resolved end to end.

File: .codex/skills/dependabot-maintainer/scripts/dependabot_maintainer.py [hunk 4]
```diff
@@ -722,7 +735,9 @@ def pyproject_extras(pyproject: Path) -> str:
 
 
 def python_bin() -> str:
-    return shutil.which("python3.12") or shutil.which("python3") or sys.executable
+    return (
+        shutil.which("python3.13") or shutil.which("python3") or sys.executable
```
[the remark this answers is not in the document]
Comment (tnorlund) [reply .codex/skills/dependabot-maintainer/scripts/dependabot_maintainer.py]: Fixed in `a2201ddde` by validating the interpreter itself before any verification venv is created:

- The maintainer probes the actual semantic version of `python3.13`, `python3`, and `sys.executable` instead of trusting executable names.
- Missing, broken, or pre-3.13 candidates are rejected while valid fallbacks remain usable.
- If no Python >=3.13 interpreter exists, `verify` now returns a clean refusal with each rejected candidate and installation guidance, and its temporary worktree cleanup still runs.
- Added five regression tests for misleading executable names, broken probes, valid fallbacks, pre-3.13 rejection, malformed probe output, and clean command failure.

Validation on Python 3.13.12: the real resolver selected `/opt/homebrew/bin/python3.13`; the two review suites passed 52/52; the CI-equivalent root/script suite passed 829 with 1 skipped; and GitHub Actions run `31532884655` is fully green across every Python package, repository tests, Lambda syntax/version checks, browser tests, and TypeScript.

File: scripts/check_python_version_consistency.py [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+        errors.append(
+            f"{relative}: {tool_name} is {value!r}; expected {expected!r}"
+        )
+
+
+def _check_pyprojects() -> list[str]:
+    errors: list[str] = []
+    for path in sorted(REPOSITORY_ROOT.rglob("pyproject.toml")):
+        if any(part.startswith(".venv") for part in path.parts):
+            continue
```
[the remark this answers is not in the document]
Comment (tnorlund) [reply scripts/check_python_version_consistency.py:170]: Fixed in `a2201ddde` by applying the shared ignored-path predicate before opening any discovered `pyproject.toml`:

- `_check_pyprojects()` now excludes `node_modules`, configured cache/generated directories, and every `.venv*` path consistently with the documentation scan.
- Regression cases place deliberately malformed TOML under every ignored directory, proving those third-party files are never parsed.
- Separate controls prove maintained projects are still validated: wrong `requires-python`, classifiers, and Black targets are reported, while malformed maintained TOML still raises instead of being hidden.

Validation on Python 3.13.12: the repository consistency checker passes; the two review suites passed 52/52; the CI-equivalent root/script suite passed 829 with 1 skipped; and GitHub Actions run `31532884655` is fully green, including the Lambda version-consistency gate and repository tests.
